contracts/utils/UniversalResolver.sol

```diff
@@ -0,0 +1,124 @@
+                address(resolver),
+                data,
+                UniversalResolver.resolveCallback.selector
+            );
+        }
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns(bool) {
+        return interfaceId == type(IExtendedResolver).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    function callWithOffchainLookupPropagation(address target, bytes memory data, bytes4 callbackFunction) internal view returns(bytes memory ret) {
+        uint256 result;
+        uint256 size;
+        
+        assembly {
+            result := staticcall(gas(), target, add(data, 32), mload(data), 0, 0)
+            size := returndatasize()
+        }
+
+        if(result == 1) {
+            // Success
+            ret = new bytes(size);
+            assembly {
+                returndatacopy(add(ret, 32), 0, size)
+            }
+            return ret;
+        }
+
+        // Failure
+        if(size >= 4) {
+            bytes4 errorId;
+            assembly {
+                returndatacopy(0, 0, 4)
+                errorId := mload(0)
+            }
+            if(errorId == OffchainLookup.selector) {
+                // Offchain lookup. Decode the revert message and create our own that nests it.
+                bytes memory revertData = new bytes(size);
+                assembly {
+                    returndatacopy(add(revertData, 32), 4, sub(size, 4))
+                }
+                (address sender, string[] memory urls, bytes memory callData, bytes4 innerCallbackFunction, bytes memory extraData) = abi.decode(revertData, (address,string[],bytes,bytes4,bytes));
+                if(sender == target) {
+                    revert OffchainLookup(address(this), urls, callData, callbackFunction, abi.encode(sender, innerCallbackFunction, extraData));
+                }
+            }
+        }
+
+        assembly {
+            returndatacopy(0, 0, size)
+            revert(0, size)
+        }
+    }
+
+    function resolveCallback(bytes calldata response, bytes calldata extraData) external view returns(bytes memory) {
+        (address target, bytes4 innerCallbackFunction, bytes memory innerExtraData) = abi.decode(extraData, (address, bytes4, bytes));
+        return abi.decode(target.functionStaticCall(abi.encodeWithSelector(innerCallbackFunction, response, innerExtraData)), (bytes));
+    }
+
+    /**
+     * @dev Finds a resolver by recursively querying the registry, starting at the longest name and progressively
+     *      removing labels until it finds a result.
+     * @return The Resolver responsible for this name, and the namehash of the full name.
+     */
+    function findResolver(bytes calldata name) public view returns(Resolver) {
+        (address resolver,) = findResolver(name, 0);
+        return Resolver(resolver);
+    }
+
+    function findResolver(bytes calldata name, uint256 offset) internal view returns(address, bytes32) {
```

Comment:
better to avoid duplicate names with l104 (maybe `findResolverWithOffset`)?

Reply:
I think this is clearer as-is.

Reply:
It's also a minor inconvenience that the frontend has to call `resolver['findResolver(bytes calldata name, uint256 offset)']` instead of just `resolver.findResolver()`. I understand that `.addr` has the similar issue but it was added due to extend the function which was unavoidable.

Reply:
This won't affect the frontend, because the second method is internal rather than public.